hnn_core/network.py

```diff
@@ -219,47 +245,86 @@ def __repr__(self):
                  len(self.pos_dict['L5_basket'])))
         return '<%s | %s>' % (class_name, s)
 
-    def _add_external_feeds(self):
-        """Legacy function, for backward compatibility with original HNN-GUI
+    def attach_drive(self, drive_name, drive_params=None,
+                     cell_specific=False):
+        """API for attaching drives according to HNN nomenclature
 
-        This function is called exactly once during initialisation
         """
-        # params of common external feeds inputs in p_common
-        # Global number of external inputs ... automatic counting
-        # makes more sense
-        # p_unique represent ext inputs that are going to go to each cell
-        self._p_common, self._p_unique = create_pext(self.params,
-                                                     self.params['tstop'])
-        self._n_common_feeds = len(self._p_common)
+        # XXX CHECKS
+        # 1. drive_name not already present
 
         # 'position' the artificial cells arbitrarily in the origin of the
         # network grid. Non-biophysical cell placement is irrelevant
         # However, they must be present to be included in gid_ranges!
         origin = self.pos_dict['origin']
 
-        # COMMON FEEDS
-        self.pos_dict['common'] = [origin for i in range(self._n_common_feeds)]
-
-        # UNIQUE FEEDS
-        for key in self._p_unique.keys():
-            # create the pos_dict for all the sources
-            self.pos_dict[key] = [origin for i in range(self.n_cells)]
+        if cell_specific:
+            drive_cell_range = range(self.n_cells)
+        else:
+            drive_cell_range = range(len(drive_params))
+        # create the pos_dict for all the sources
+        self.pos_dict[drive_name] = [origin for i in drive_cell_range]
 
-        # Now add the names of the feeds to a list
-        self.feedname_list.append('common')
-        # grab the keys for the unique set of inputs and sort the names
-        # append them to the src list along with the number of cells
-        unique_keys = sorted(self._p_unique.keys())
-        self.feedname_list += unique_keys
+        # NB must be added before updating gid_ranges!
+        self.external_drives.update({drive_name: list()})
 
         # Every time pos_dict is updated, gid_ranges must be updated too
         self._update_gid_ranges()
 
-        # Create the feed dynamics (event_times)
-        self._instantiate_feeds(n_trials=self.params['N_trials'])
-
-    def _instantiate_feeds(self, n_trials=1):
-        """Creates event_time vectors for all feeds and all trials
+        drive_dict = {'drive_name': drive_name, 'params': drive_params,
```

Comment:
I think we should avoid creating more 'params' (sub)dicts. What if we instead dismantled the fundamental components of `p_common` and `p_unique` so that we instead have a `syn_weights` key that contains a dict value (we'll allow this to be a dict because we shouldn't hardcode the cell/network structure) and single value keys associated with `t_start`, `t_stop`, etc.

Reply:
We could also add a key called `renewal_process` that determines the rhythmicity/Poisson-ness for the duration of the drive. If the drive happens to have `t_start`== `t_stop` or `t_stop`==`None`, than it is an evoked drive with no duration regardless of `renewal_process`. If `t_start`<`t_stop` and `renewal_process`==`rhythmic`, the input is rhythmic. I'm getting way ahead of myself, but in the future we could even use a single float to define the Fano factor of the time events defined by renewal process.

Reply:
I don't think we can avoid a `dict` data structure so easily. The reason is that the connections are instantiated by `Neuron` at a later time than when specified by the user. The trick is to have a transparent object -- clearly defined, easy to inspect, not too nested and simple. It's why I suggested repurposing `self.feed_times` in #224